crates/pathfinder/src/sequencer.rs

```diff
@@ -121,37 +121,82 @@ async fn parse_raw(resp: reqwest::Response) -> Result<reqwest::Response, Sequenc
 
 /// Wrapper function to allow retrying sequencer queries in an exponential manner.
 ///
-/// Initial backoff time is 2 seconds. Retrying stops after approximately 4 minutes in total.
+/// Retry is performed on __all__ types of errors __except for__
+/// [StarkNet specific errors](crate::sequencer::error::StarknetError).
+///
+/// Initial backoff time is 30 seconds and saturates at 1 hour:
+///
+/// `backoff [secs] = min((2 ^ N) * 15, 3600) [secs]`
+///
+/// where `N` is the consecutive retry iteration number `{1, 2, ...}`.
 async fn retry<T, Fut, FutureFactory>(future_factory: FutureFactory) -> Result<T, SequencerError>
 where
     Fut: Future<Output = Result<T, SequencerError>>,
     FutureFactory: FnMut() -> Fut,
 {
     use crate::retry::Retry;
     use reqwest::StatusCode;
-    use std::num::{NonZeroU64, NonZeroUsize};
+    use std::num::NonZeroU64;
+    use tracing::{info, trace, warn, Level};
 
     Retry::exponential(future_factory, NonZeroU64::new(2).unwrap())
-        // Max number of retries of 7 gives a total accumulated timeout of 4 minutes and 15 seconds (2^8-1)
-        .max_num_retries(NonZeroUsize::new(7).unwrap())
+        .factor(NonZeroU64::new(15).unwrap())
+        .max_delay(Duration::from_secs(60 * 60))
         .when(|e| match e {
-            SequencerError::TransportError(te) if te.is_timeout() => {
-                tracing::debug!("Retrying due to timeout");
+            SequencerError::ReqwestError(e) => {
+                match tracing::level_filters::LevelFilter::current().into_level() {
+                    None => {}
+                    Some(level) if level >= Level::TRACE => {
+                        trace!("Retrying due to: {:?}", e);
+                    }
+                    Some(_) => {
+                        if e.is_timeout() {
+                            info!("Retrying due to a timeout.");
+                        } else if e.is_decode() {
+                            warn!("Retrying due to a deserialization error.");
```

Comment:
Should this maybe be fatal? I think at least `error!` level if we stick to logs.

Reply:
Do we care for transient errors in the sequencer output?
If yes, this should be an error I guess, but we should maybe continue retrying.
However, this also means that in case Starkware adds a new attribute to one of their JSON objects (highly likely on Starknet alpha I guess) we'll still have a lot of pathfinder instances polling the API in vain.

Reply:
> we'll still have a lot of pathfinder instances polling the API in vain.

Hence the backoff quickly saturating at 1 hr to decrease the strain on the API.
Whether this should be a fatal error - I am not sure, at the moment I am assuming a more favorable [scenario first - that it's a bug and will be fixed](https://github.com/eqlabs/pathfinder/issues/246#issuecomment-1105125724).
Right now the log would just indicate that L2 is not progressing and in case this is an update the update monitoring would indicate a need for update as well without stopping the process.

Unfortunately quite a lot of the instances of pathfinder are wrapped in a systemd unit which does not stop upon a fatal but blindly restarts the service:
```
[Service]
User=$USER
Type=simple
WorkingDirectory=$HOME/pathfinder/py
ExecStart=/bin/bash -c \"source $HOME/pathfinder/py/.venv/bin/activate && /usr/local/bin/pathfinder --http-rpc=\"0.0.0.0:9545\" --ethereum.url $ALCHEMY\"
Restart=on-failure
LimitNOFILE=65535
```

Reply:
Fair point. I think retry always is good. I forgot about the service wrapping; which would just put us back in the same position we are in.

But definitely upgrade the level to `error!` for cases we really want to know about.

Reply:
Re systemd: it's true that systemd restarts pathfinder, but it also has a rate-limit for starting a unit: https://www.freedesktop.org/software/systemd/man/systemd.unit.html#StartLimitIntervalSec=interval
The defaults are 10s and 5 (5 starts within 10s), which are probably not reached even if we panic right after receiving the first JSON response.

I think you're right, we could just keep with `error!` and the slow retries.